EIPS/eip-4573.md

```diff
@@ -0,0 +1,138 @@
+eip: 4573
+title: Entry Points and Procedures for EVM Code Sections
+description: This proposal introduces EVM procedures as entry points to EOF code sections.
+status: Draft
+type: Standards Track
+category: Core
+author: Greg Colvin <greg@colvin.org>
+discussions-to: https://ethereum-magicians.org/t/eip-4573-entry-points-for-eof/4573
+created: 
+requires: 2315, 3540, 3670, 3779, 4200
+---
+
+## Abstract
+
+EOF sections and EVM instructions are introduced to provide procedural entry points to code sections.
+
+## Motivation
+
+Currently the Ethereum Object Format supports a single code section, with no further means of structuring the code.
+
+This proposal adds a layer of structure by supporting additional code sections.  How to access these sections?  This proposal specfies a table of entry points for each additional section.  Each entry point corresponds to a procedure within a code section.
+
+We define procedures as blocks of code that can be entered only at their entry point, and at other points can call other procedures and subroutines and return to the procedure that called them.
+
+## Specification
+
+### EOF Sections
+
+Code sections after the first must be immediately preceded by an entry point section.  These sections can only be entered at one of the specified entry points in the code section, given as an offset relative to the beginning of the code section.
+
+#### Entry Point Section
+
+| description| | length  ||
+|------------|--|---------|-|
+| section_kind|  | 1-byte  | Encoded as a 8-bit unsigned number. |
+| section_size | | 2-bytes | Encoded as a 16-bit unsigned big-endian number. |
+| []entry_point | |
+| | entry_point  | 2-bytes | Encoded as a 16-bit unsigned big-endian number. |
+| | ... | ... | ... |
+
+### Call Frame Stack
+
+These instructions make use of a `frame_stack` to hold arguments, locals, and return values for procedures in memory.  Frame memory begins at address 0 in memory and grows downwards, towards more negative addresses.
+
+Memory can be addressed relative to the frame point `FP` or  by absolute address.  `FP` starts at 0, and moves to point to the start of the input arguments for each `CALLPROC`, and to the start of the output values for each `RETURNPROC`.
+
+So on entry to a procedure the call frame stack looks like this `(FP <= 0)`:
+
+```
+     0-> . . .
+         . . .
+         previous frame
+         . . .
+    FP-> inputs
+         . . .
+         locals
+         . . .
+```
+And on exit from the procedure the call frame stack looks like this:
+
+```
+     0-> . . .
+         . . .
+    FP-> previous frame
+         . . .
+         outputs
+         . . .
+```
+
+*Note:  In order to save space the inputs and outputs overlap.  When the inputs are used up in the course of producing the outputs this should not be a problem.  (The space beyond the end of the call frame can be used for temporaries.)*
+
+### Instructions
+
+#### ENTERPROC (0x??) n_inputs: uint2, n_locals: uint2, n_outputs: uint2
```

Comment:
s/uint2/uint256/?  Surely you expect more than 3 inputs, outputs and locals.

Reply:
I better  add more description ;)  Those  are the counts of how many inputs, locals, and outputs are allocated.   The way I use them elsewhere is consistent with this.